pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -60,128 +64,175 @@
-              // Only update the valid doc ids for the new segment
-              if (validDocIds == currentRecordLocation.getValidDocIds()) {
-                validDocIds.remove(currentRecordLocation.getDocId());
-              }
+              validDocIds.remove(currentRecordLocation.getDocId());
               validDocIds.add(recordInfo._docId);
-              return new RecordLocation(segmentName, recordInfo._docId, recordInfo._timestamp, validDocIds);
+              return new RecordLocation(immutableSegment, recordInfo._docId, recordInfo._timestamp);
             } else {
               return currentRecordLocation;
             }
-          } else {
-            // The current record location is pointing to a different segment
-
-            // Update the record location when getting a newer timestamp, or the timestamp is the same as the current
-            // timestamp, but the segment has a larger sequence number (the segment is newer than the current segment).
-            if (recordInfo._timestamp > currentRecordLocation.getTimestamp() || (
-                recordInfo._timestamp == currentRecordLocation.getTimestamp()
-                    && LLCSegmentName.isLowLevelConsumerSegmentName(segmentName)
-                    && LLCSegmentName.isLowLevelConsumerSegmentName(currentRecordLocation.getSegmentName())
-                    && LLCSegmentName.getSequenceNumber(segmentName) > LLCSegmentName
-                    .getSequenceNumber(currentRecordLocation.getSegmentName()))) {
-              currentRecordLocation.getValidDocIds().remove(currentRecordLocation.getDocId());
+          }
+
+          // The current record is in an old segment being replaced
+          // This could happen when committing a consuming segment, or reloading a completed segment. In this case, we
+          // want to update the record location when there is a tie because the record locations should point to the new
+          // added segment instead of the old segment being replaced. Also, do not update the valid doc ids for the old
+          // segment because it has not been replaced yet.
+          String currentSegmentName = currentSegment.getSegmentName();
+          if (segmentName.equals(currentSegmentName)) {
+            if (recordInfo._timestamp >= currentRecordLocation.getTimestamp()) {
               validDocIds.add(recordInfo._docId);
-              return new RecordLocation(segmentName, recordInfo._docId, recordInfo._timestamp, validDocIds);
+              return new RecordLocation(immutableSegment, recordInfo._docId, recordInfo._timestamp);
             } else {
               return currentRecordLocation;
             }
           }
+
+          // The current record is in a different segment
+          // Update the record location when getting a newer timestamp, or the timestamp is the same as the current
+          // timestamp, but the segment has a larger sequence number (the segment is newer than the current segment).
```

Comment:
why do we need this additional handling and split the handling of immutable segment vs consuming segment? Why does partial upsert require this change?

Reply:
Partial upsert requires storing the actual segment instead of the segment name because it requires reading the previous record from the segment.
The handling logic is the same as full upsert, where the immutable segment (completed segment) is loaded via the `addSegment()` (per-segment load), and the consuming segment is loaded via the `updateRecord()` (per-record load).